#VRML_SIM R2023a utf8
# license: Copyright Cyberbotics Ltd. Licensed for use only with Webots.
# license url: https://cyberbotics.com/webots_assets_license
# documentation url: https://www.cyberbotics.com/doc/guide/lidar-sensors#lds-01
# Model of the Robotis LDS-01 Laser Distance Sensor
# Reference: http://www.robotis.us/360-laser-distance-sensor-lds-01-lidar/
# template language: javascript

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/appearances/protos/RoughPolymer.proto"

PROTO RobotisLds01 [
  field SFVec3f    translation       0 0 0.02
  field SFRotation rotation          0 0 1 0
  field SFString   name              "LDS-01"
  field SFFloat    noise             0.0043
  field SFBool     enablePhysics     TRUE
]
{
  Lidar {
    translation IS translation
    rotation IS rotation
    children [
      Transform {
        rotation 0 0.707103 0.70711 3.141585
        children [
          HingeJoint {
            jointParameters HingeJointParameters {
              axis 0 1 0
            }
            device [
              RotationalMotor {
                name %<= '"' + fields.name.value + '_main_motor"' >%
                maxVelocity 40
              }
            ]
            endPoint Solid {
              translation 0 -0.01 0
              rotation 1 0 0 -1.5708
              scale 0.001 0.001 0.001
              children [
                Shape {
                  appearance DEF MAIN_APPEARANCE RoughPolymer {
                    baseColor 0 0 0
                    textureTransform TextureTransform {
                      scale 4 4
                    }
                  }
                  geometry Mesh {
                    url "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/devices/robotis/protos/meshes/turret.obj"
                  }
                }
              ]
              name "lower"
            }
          }
          HingeJoint {
            jointParameters HingeJointParameters {
              axis 0 1 0
              anchor -0.046 0 0
            }
            device [
              RotationalMotor {
                name %<= '"' + fields.name.value + '_secondary_motor"' >%
                maxVelocity 150
              }
            ]
            endPoint Solid {
              translation 0 -0.01 0
              rotation 1 0 0 -1.5708
              scale 0.001 0.001 0.001
              children [
                Shape {
                  appearance PBRAppearance {
                    baseColor 0.886671 0.833646 0.78027
                    metalness 0
                  }
                  geometry Mesh {
                    url "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/devices/robotis/protos/meshes/wheel.obj"
                  }
                }
              ]
            }
          }
          Transform {
            translation 0 -0.01 0
            rotation 1 0 0 -1.5708
            children [
              Shape {
                appearance PBRAppearance {
                  baseColor 0 0 0
                  roughness 1
                  metalness 0
                }
                geometry Mesh {
                  url "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/devices/robotis/protos/meshes/belt.obj"
                }
              }
            ]
          }
          Transform {
            translation 0 -0.01 0
            rotation -1 0 0 1.5708
            scale 0.001 0.001 0.001
            children [
              Shape {
                appearance DEF MAIN_APPEARANCE RoughPolymer {
                  baseColor 0 0 0
                  textureTransform TextureTransform {
                    scale 4 4
                  }
                }
                geometry Mesh {
                  url "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/devices/robotis/protos/meshes/base.obj"
                }
              }
            ]
          }
        ]
      }
    ]
    name IS name
    boundingObject Group {
      children [
        Transform {
          translation 0.014 0 -0.014
          rotation -0.577352 0.577348 0.577352 -2.094405
          children [
            Box {
              size 0.07 0.01 0.1
            }
          ]
        }
        Transform {
          translation 0 0 0
          children [
            Cylinder {
              height 0.021
              radius 0.032
            }
          ]
        }
      ]
    }
    %< if (fields.enablePhysics.value) { >%
    physics Physics {
      density -1
      mass 0.125
    }
    %< } >%
    horizontalResolution 60
    fieldOfView 3.14159
    numberOfLayers 1
    near 0.07
    minRange 0.1
    maxRange 4
    noise IS noise
  }
}
